feat(cli/deploy): add db snapshot flag for simple-rds [INFENG-269]

loksonarius opened the pull request:
## Description

<!---
Lead with the intended commit body in this description field. For breaking
changes, please include "BREAKING CHANGE:" at the beginning of your commit
body.  At a minimum, this section should include a bracketed reference to the
Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly
into the description field.
-->

Adds a `--db-snapshot` flag to provide a DB snapshot to initialize the RDS instance with for the Simple RDS deployment type.


## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot
as appropriate.  Reviewers may ask questions about this test plan to ensure
adequate manual coverage of changes.
-->

Local testing of added template logic done -- the username/password parameters are mutually exclusive with the snapshot parameter.

## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for
particularly tricky bits of code that could use extra scrutiny, historical
context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->

This change is made not to fill a user request but to enable the performance testing suites to boot instances pre-configured with the data necessary to run tests instead of having to manually load the data after boot.

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
<!---
Retain the relevant line and replace 000 with ticket number.

DET-000
MLG-000
WEB-000
DESIGN-000
No Ticket
--->

INFENG-269

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

-->

stoksc commented on harness/determined/deploy/aws/templates/simple-rds.yaml:
```diff
@@ -469,9 +474,13 @@ Resources:
       Engine: postgres
-      MasterUsername: postgres
-      MasterUserPassword: !Ref DBPassword
+      MasterUsername:
+        !If [NoDBSnapshot, "postgres", !Ref "AWS::NoValue"]
+      MasterUserPassword:
+        !If [NoDBSnapshot, !Ref DBPassword, !Ref "AWS::NoValue"]
```
Will the user need to pass something here? Otherwise there will be no value in the master config and it would fail to connect?

loksonarius replied:
In my testing, this just caused the DB to generate a password on setup.
That's AWS behavior, so I'm not sure if that's the intended behavior for `det deploy`, but it's pretty sensible.

ioga replied:
but how would the det master know the generated password?

stoksc replied:
yeah, ^ my thoughts exactly

loksonarius replied:
So I see how the problem would manifest if we were providing the template to a user to deploy manually, but I'm not sure how that affects the setup here.
The password flag isn't currently required since it's set to some default, and things deploy using that default currently. This portion of the template doesn't affect if/how the password is rendered into any `master.yml` config file -- is a password is passed, it'll still be in the master config generated, if it isn't then the master config is still generated with the same default password.

I'm not sure I understand if the question is about what I've implemented or about `det deploy`'s behavior in general rn

stoksc replied:
I think we should just remove this and configure the database with a password still. I've tested it, it works fine and it's much generally useful. It's much easier to use than needing to know the database password for the snapshot.

I tested this out and immediately hit this because the database snapshot I used did not use the default password.

stoksc replied:
Even after what we discussed, this is still going to make it so it only ever works with the default password since we've just made it impossible to specify what the any password other than the default. I think if we are saying "DBPassword isn't allowed to be set on the rds" that is fine but doesn't a user still need to pass `--db-password` to let us know what the snapshot password is?

loksonarius commented on harness/determined/deploy/aws/cli.py:
```diff
@@ -172,6 +172,13 @@ def deploy_aws(command: str, args: argparse.Namespace) -> None:
 
+        if args.db_snapshot is not None and args.db_password != constants.defaults.DB_PASSWORD:
+            # this precedence is implemented in the deployment template
+            print(
+                "--db-snapshot and --db-password are mutually exclusive, "
+                "password value be ignored as snapshot would take precedence"
+            )
```
It's not an error though, but just a heads up.
Or are you suggesting I explicitly make it a failure?

loksonarius replied:
So the db password is a required flag if the password is non-default. In second thought, I'm not sure this heads up is actually an error we can enforce

stoksc replied:
I think that's exactly my point with the other thread; I don't think they should even be mutually exclusive.

stoksc commented:
Let's just sync offline about this PR

loksonarius commented:
Discussed this PR and the dependent work with @stoksc today.
Going to put this PR on hold (still addressing feedback on incorrect defaults and info messages) until I finish modifying our performance testing code to use a locally built CLI and latest published image.

dannysauer commented on .github/workflows/performance-tests.yml:
```diff
@@ -30,19 +30,33 @@ jobs:
+          git rev-parse HEAD > .master-image-version
+          echo "Looking for master image version $(cat .master-image-version)"
+          docker pull "determinedai/determined-master:$(cat .master-image-version)"
+          echo "text=$(cat .master-image-version)" >> "$GITHUB_OUTPUT"
```
If you're only ever going to use this value in steps within the same job, I'd probably make it easier to use by putting it as a variable in GITHUB_ENV instead of setting it as step output. :)  Up to you, though.

dannysauer reviewed (approved):
Infra-approved

stoksc reviewed (approved):
lgtm.

separate pr to add `--db-snapshot`?

dzhu commented on docs/release-notes/db-snapshot-flag.rst:
```diff
@@ -0,0 +1,7 @@
+
+**New Features**
+
+-  CLI: Add new `--db-snapshot` flag for the `det deploy aws up` subcommand that allows starting RDS
+   Database instances with a pre-existing snapshot. This flag is only currently usable with the
+   `simple-rds` deployment type.
```
```suggestion
-  CLI: Add new ``--db-snapshot`` flag for the ``det deploy aws up`` subcommand that allows starting RDS
   DB instances with a pre-existing snapshot. This flag is only currently usable with the
   ``simple-rds`` deployment type.
```
(From a quick look at the RDS docs, it seems like they very consistently use "DB instance" for this.)

dzhu commented on harness/determined/deploy/aws/cli.py:
```diff
@@ -168,10 +168,22 @@ def deploy_aws(command: str, args: argparse.Namespace) -> None:
+            # this precedence is implemented in the deployment template
+            print(
+                "--db-snapshot and --db-password are mutually exclusive, "
+                "password value be ignored as snapshot would take precedence"
```
```suggestion
                "password value ignored as snapshot would take precedence"
```

stoksc commented:
> separate pr to add --db-snapshot?
separate pr to add --db-snapshot ...arn... to CI

i was even confused by the ambiguity of my comment

daranday commented on harness/determined/deploy/aws/cli.py:
```diff
@@ -172,10 +172,22 @@ def deploy_aws(command: str, args: argparse.Namespace) -> None:
 
+        if args.db_snapshot is not None and args.db_password != constants.defaults.DB_PASSWORD:
+            # this precedence is implemented in the deployment template
+            print(
```
Mutually exclusive sounds like it should be an error. Perhaps we can just mention the precedence?
